### PR TITLE
Added delegate call mechanism for contracts to swap using meta-aggregator without approval.

### PR DESCRIPTION
**Pull request by jaysinghdhakad**
The following method are made for contract to swap using meta-aggregator swap contract. 

1. swapETHDelegate: This method can be used by contracts to swap ETH to tokens
2. swapERC20Delegate: This method can be used by contracts to swap tokens. The contract don't need to give apporval for its swap.

### CODE REVIEW

File: contracts/meta-aggregator/MetaAggregatorSwapContract.sol [hunk 1]
```diff
@@ -120,23 +150,68 @@ contract MetaAggregatorSwapContract is
...
+    function swapETHDelegate(
+        address tokenIn,
+        IERC20 tokenOut,
+        address aggregator,
+        bytes calldata swapData,
+        uint256 amountIn,
+        uint256 minAmountOut,
+        address receiver,
+        bool isDelegate
+    ) external payable returns (uint256 amountOut) {
```
Comment (langnavina97) [inline contracts/meta-aggregator/MetaAggregatorSwapContract.sol:178]: Add `nonReentrant`

File: contracts/meta-aggregator/MetaAggregatorSwapContract.sol [hunk 2]
```diff
@@ -145,49 +220,106 @@ contract MetaAggregatorSwapContract is
...
-        if (amountIn == 0) {
-            revert AmountInMustBeGreaterThanZero();
-        }
-        if (minAmountOut == 0) {
-            revert MinAmountOutMustBeGreaterThanZero();
-        }
-        if (address(tokenIn) == address(tokenOut)) {
-            revert TokenInAndTokenOutCannotBeSame();
-        }
+    ) external returns (uint256 amountOut) {
```
Comment (langnavina97) [inline contracts/meta-aggregator/MetaAggregatorSwapContract.sol:223]: Add `nonReentrant`

File: contracts/meta-aggregator/MetaAggregatorSwapContract.sol [hunk 3]
```diff
@@ -120,23 +150,68 @@ contract MetaAggregatorSwapContract is
...
+     * @dev Swaps ETH using delegatecall.
+     * This function is called when the swap needs to happen in a delegatecall context.
+     * @param tokenIn must be the native token.
+     * @param tokenOut The ERC20 token to swap to.
+     * @param aggregator The address of the aggregator to use for the swap.
+     * @param swapData The data required for the swap.
+     * @param amountIn The amount of ETH to swap.
+     * @param minAmountOut The minimum amount of tokenOut expected.
+     * @param receiver The address to receive the tokenOut.
+     * @param isDelegate Indicates if the swap is in a delegatecall context.
```
Comment (langnavina97) [inline contracts/meta-aggregator/MetaAggregatorSwapContract.sol:166]: Any specific reason you added the parameter `isDelegate`? The function description says "Swaps ETH using delegatecall."

File: contracts/meta-aggregator/MetaAggregatorSwapContract.sol (hunk 1 above)
Comment (jaysinghdhakad) [reply contracts/meta-aggregator/MetaAggregatorSwapContract.sol:178]: This one is tricky would the calling contract have nonReentrant in its contract. Would the storage place for the nonReentrant status variable aline with our contract storage. That is why I didn't use nonReentrant modifier here.

File: contracts/meta-aggregator/MetaAggregatorSwapContract.sol (hunk 2 above)
Comment (jaysinghdhakad) [reply contracts/meta-aggregator/MetaAggregatorSwapContract.sol:223]: Same as above.

File: contracts/meta-aggregator/MetaAggregatorSwapContract.sol (hunk 3 above)
Comment (jaysinghdhakad) [reply contracts/meta-aggregator/MetaAggregatorSwapContract.sol:166]: It is for the aggregator called by our contract. If enso this is true or others this is false.

**Comment by Havoc19** [comment]
We don't need here swapETHDelegate as swapETH, does the same thing and also has a isDelegate bool, if I delegate call swapETH it should work

**Comment by jaysinghdhakad** [comment]
> We don't need here swapETHDelegate as swapETH, does the same thing and also has a isDelegate bool, if I delegate call swapETH it should work @jaysinghdhakad

It runs in the context of the calling contract so we don't need to transfer any token if the receiver is same as the calling contract.

**Comment by jaysinghdhakad** [comment]
> > > We don't need here swapETHDelegate as swapETH, does the same thing and also has a isDelegate bool, if I delegate call swapETH it should work @jaysinghdhakad
> > 
> > 
> > It runs in the context of the calling contract so we don't need to transfer any token if the receiver is same as the calling contract. Also enso gives two different address for swapping eth 0x80EbA3855878739F4710233A8a19d89Bdd2ffB8E(for routing) and 0x38147794FF247e5Fc179eDbAE6C37fff88f68C521(for delegate) i.e two different methods. @Havoc19
> 
> enso gives 2 different address for swapping eth, what addresses are those swap_target? But we can delegate call swapEth and swapERC20, right? and the same thing we are doing for delegateEth and delegateERC20

The difference is the transfer of tokens. We won't need to transfer tokens in case the sender and receiver are not different for a delegate call which would save gas. if we integrate both delegate and simple call logic to the same function the gas cost for the swap would increase. With swapERC20 and swapERC20 delegate the simple call has to be done through the manager contract not the swap contract so that the the vulnerability of  draining of funds would not happen for delegate call we don't need the manager contract for swap. It could be directly done on swap contract. This would save gas for contract calls. Other reason to do this is not to increase the complexity so some bug would creap in.

**Comment by langnavina97** [comment]
Code changes look good. Can you please add some test cases for different amounts (low and high amounts) to test the behavior? And you can also include different protocols that we're integrating to make sure it doesn't only work for base tokens..

**Comment by jaysinghdhakad** [comment]
> Code changes look good. Can you please add some test cases for different amounts (low and high amounts) to test the behavior? And you can also include different protocols that we're integrating to make sure it doesn't only work for base tokens..

What protocols do you want me test ?

**Comment by langnavina97** [comment]
@jaysinghdhakad are you pushing the changes for the reentrancy check?

**Review by Havoc19 — APPROVED** [review]

**Review by langnavina97 — APPROVED** [review]

**Review by aj07 — APPROVED** [review]

**Pull request merged**